### ISSUE TITLE
Visualization does not handle updates to a source file properly

### ISSUE THREAD

**Issue by skogsbaer** [skogsbaer/write-your-python-program#163]
Given this program:

```python
def incByOneInPlace(l: list[int]) -> None:
    for i in range(len(l)):
        l[i] = l[i] + 1

x = [5, 3, 2, 0]
incByOneInPlace(x[:]) # swap x and x[:]
```

In todays lecture, I replaced `x[:]` with `x` and back. But the visualization sometimes displays this

<img width="732" alt="Bildschirmfoto 2024-12-02 um 09 38 01" src="https://github.com/user-attachments/assets/6909cb4c-1bdc-4989-b3f5-01beb366434c">

even though the file contains `x` and it sometimes displays 

<img width="734" alt="Bildschirmfoto 2024-12-02 um 09 39 07" src="https://github.com/user-attachments/assets/cc3c62d6-77ca-4bdf-9d3d-00b1589f4976">

although I wrote `x[:]`

I could not determine a pattern for this behavior.

**Comment by hannesbraun** [comment]
As far as I can tell, two lists will be displayed for `x[:]`, one list will be displayed for `x`. The reason for this is that `x` and `x[:]` have different identities.

```python
x = [1, 2, 3]
print(id(x))
print(id(x[:]))
```

**Comment by skogsbaer** [comment]
Yes, that's true. But in the lecture, I noticed that the visualization sometimes seems not to pick up the last changes to a file. Given the program

```python
x = [1,2,3]
y = x[:]
```

it displayed one list with two arrows pointing to it from x and y (which is wrong).

I think this was because I had the program

```python
x = [1,2,3]
y = x
```

before, and the visualization still rendered the "old" program.

Is this possible?

**Comment by hannesbraun** [comment]
This should only be possible if both versions of the code have the same MD5 hash and the version with `y = x` was executed first. Otherwise, changes are automatically saved and the trace generator re-generates the trace if none is available in the cache for the hash of the input file.

The only thing that could have also happened is that you somehow had the old visualization still open, started the new visualization and somehow landed in the old tab (for whatever reason) as it won't be closed automatically. But I'm not sure how realistic that is.

**Comment by skogsbaer** [comment]
Can we (should we?) close the old visualization tab automatically? So that at most one visualization tab is visible?

**Comment by hannesbraun** [comment]
We *can* close the old visualization tab. But I'm not sure if we should. Keeping the old tab open leaves you with the option to open two tabs side by side for comparing some steps in case someone wants to do this. I'd personally not close the tab automatically. But ultimately, I'm open to whatever you think is the better solution.

**Comment by hannesbraun** [comment]
Hmm, yes, this is true. Maybe we should close it then as soon as we open a new one...

**Comment by skogsbaer** [comment]
> Hmm, yes, this is true. Maybe we should close it then as soon as we open a new one...

Yes, I would prefer this solution.
